Add summary printing for GenericMemoryRef

vishal2005025 opened the pull request:
## Fixes #57184 

## What this PR does
Improves the summary display of `GenericMemoryRef` so that `BoundsError` messages show the effective accessible length from the current reference offset.

## Why
When a `MemoryRef` is created at a non-zero offset, the existing `BoundsError` message can be misleading because it only shows the type. Displaying the remaining accessible length makes the error more actionable and aligns the behavior with other types like `SubString`.

## Implementation details
- Adds a `Base.summary` method for `GenericMemoryRef` in `base/show.jl`
- Computes remaining length using `parent(ref)` and `Core.memoryrefoffset(ref)`
- Does not change bounds checking logic or error types
- Adds focused tests in `test/errorshow.jl` to validate the improved error output

vishal2005025 committed:
Improve BoundsError summary for GenericMemoryRef

vishal2005025 committed:
Improve BoundsError summary for GenericMemoryRef

vishal2005025 committed:
Merge branch 'feat/genericmemoryref-summary' of https://github.com/vishal2005025/julia into feat/genericmemoryref-summary

vishal2005025 committed:
Merge branch 'master' into feat/genericmemoryref-summary

vishal2005025 committed:
Fix MemoryRef BoundsError summary test expectations

vishal2005025 committed:
whitespace fix

vishal2005025 committed:
trailing new line fix

vishal2005025 commented:
@LilithHafner @adienes @KristofferC  please take a look.

LilithHafner commented on base/show.jl:
```diff
@@ -3128,6 +3128,24 @@ function array_summary(io::IO, a, inds)
+Print a brief description of a `GenericMemoryRef`, showing the remaining
+accessible length from the current offset. This improves error messages
+like `BoundsError` to reflect the effective size, similar to `SubString`.
+"""
```
There's no need to add a docstring for methods that are already documented by the generic function docstring and don't deviate from the generic interface.

```suggestion
```

LilithHafner commented on base/show.jl:
```diff
@@ -3128,6 +3128,24 @@ function array_summary(io::IO, a, inds)
+function Base.summary(io::IO, ref::GenericMemoryRef{kind,T,AS}) where {kind,T,AS}
+    mem = parent(ref)
+    idx = Core.memoryrefoffset(ref)            # 1-based offset within parent memory
+    rem = length(mem) - (idx - 1)              # remaining accessible elements
```
I prefer the variable naming in https://github.com/JuliaLang/julia/pull/57278/files

LilithHafner reviewed (commented):
This is a duplicate of https://github.com/JuliaLang/julia/pull/57278/files, but this PR includes tests, so let's try proceeding with this PR instead.

vishal2005025 committed:
Simplify GenericMemoryRef summary implementation

vishal2005025 commented:
@LilithHafner Thanks for the feedback! I’ve addressed all the review comments by simplifying the summary implementation, aligning naming with PR #57278, and removing the unused branches.

Please take a look.

vishal2005025 committed:
Use preferred GenericMemoryRef summary implementation

Co-authored-by: Omar Elrefaei <omar.elrefaei@users.noreply.github.com>

vishal2005025 commented:
@LilithHafner Thanks for the clarification! I’ve updated the implementation to match @Omar-Elrefaei’s version exactly and added him as a co-author, while keeping the tests. Please let me know if anything else is needed.
